frontend/src/metabase/common/components/CollectionPicker/components/CollectionItemList.tsx

```diff
@@ -1,29 +1,47 @@
 import { skipToken, useListCollectionItemsQuery } from "metabase/api";
+import { isNullOrUndefined } from "metabase/lib/types";
+import type { CollectionItem, DatabaseId } from "metabase-types/api";
 
 import { ItemList } from "../../EntityPicker";
 import type { CollectionItemListProps, CollectionPickerItem } from "../types";
 
+interface Props extends CollectionItemListProps {
+  databaseId?: DatabaseId;
+}
+
 export const CollectionItemList = ({
+  databaseId,
   query,
   onClick,
   selectedItem,
   isFolder,
   isCurrentLevel,
   shouldDisableItem,
-}: CollectionItemListProps) => {
+}: Props) => {
   const {
     data: collectionItems,
     error,
     isLoading,
   } = useListCollectionItemsQuery<{
-    data: { data: CollectionPickerItem[] };
+    data: {
+      data: (CollectionPickerItem & Pick<CollectionItem, "database_id">)[];
+    };
     error: any;
     isLoading: boolean;
   }>(query ? query : skipToken);
 
+  const items = collectionItems?.data ?? [];
+  const filteredItems = items.filter(item => {
+    if (isNullOrUndefined(databaseId) || isNullOrUndefined(item.database_id)) {
+      return true;
+    }
+
+    return item.database_id === databaseId;
+  });
```

Comment:
I don't feel super-strongly about this, but rather than build join-specific logic into this component, I'd rather pass a `itemFilter` function kind of like `shouldDisableItem` that would be more generally useful

Reply:
Done: 2b944f0d6ad396e2720b0e3278772b030d50cf2

`QuestionPicker` & `EntityPickerModal` still accept `databaseId` prop due to it being useful in `EntityPickerSearchInput`.